library/init/meta/interactive_base.lean

```diff
@@ -195,6 +195,33 @@ meta structure inductive_decl :=
 /-- Parses and elaborates a single or multiple mutual inductive declarations (without the `inductive` keyword), depending on `is_mutual` -/
 meta constant inductive_decl.parse : decl_meta_info → parser inductive_decl
 
+universes u
+
+/-- Typeclass for custom interaction monads, which provides
+    the information required to convert an interactive-mode
+    construction to a `tactic` which can actually be executed.
+
+    Given a `[monad m]`, `execute` explains how to turn a `begin ... end`
+    block, or a `by ...` statement into a `tactic α` which can actually be
+    executed. The `execute_with` function facilitates the passing of an
+    optional configuration parameter `config`, using the syntax:
+    ```
+    begin [custom_monad] with config,
+        ...
+    end
+    ```
+-/
+meta class executor (m : Type → Type u) [monad m] :=
+(execute : m unit → tactic unit)
+(config_type : Type := unit)
+(execute_with : config_type → m unit → tactic unit := λ _, execute)
```

Comment:
How about having only `execute_with`, and provide a default `config_type` object to pass? That seems consistent with the smt example, and with most use cases I can think of.

Reply:
Sounds great. Done.